src/canvas/Exporter.tsx

```diff
@@ -0,0 +1,173 @@
+            new PatternPath(1, [new LineSegment(new Point(142, 172), new Point(223, 222))]),
+            new PatternPath(1, [new LineSegment(new Point(223, 222), new Point(95, 326))]),
+        ];
+        testPiece = new PatternPiece(paths, AllowanceFinder.computeAllowancePaths(paths, allowanceMapTestBig));
+        this._testPiecesBig = [testPiece];
+    }
+
+    save = (): void => {
+        this.doc = new jsPDF('p', 'in', 'letter');
+        // Delete the first page, which is automatically added when a new jsPDF is created.
+        this.doc.deletePage(1);
+
+        let patternPieces = this._documentModel.getPatternPieces();
+        
+        // TODO remove, this is for testing.
+        if (!patternPieces?.length) {
+            patternPieces = this._testPiecesBig;
+        }
+
+        console.log("pattern pieces:", patternPieces);
+
+        // Set page height, width and DPI
+        const DPI = 300;
+        const pageWidth = 8.5;
+        const pageHeight = 11;
+        const margin = .5;
+        const innerPageWidth = pageWidth - (2 * margin);
+        const innerPageHeight = pageHeight - (2 * margin);
+        const pageSizeX = innerPageWidth * DPI;
+        const pageSizeY = innerPageHeight * DPI;
+
+        /* 
+        * Create a canvas, which we will draw to. Then we will turn that canvas into a png 
+        * and add it to the pdf.
+        */
+        const canvas = this._createCanvas(DPI, innerPageWidth, innerPageHeight);
+        const ctx = canvas.getContext('2d');
+
+        // Calculate ratio to scale by.
+        const pixelsPerInch = this._documentModel.getSizeRatio() > 0 ? this._documentModel.getSizeRatio() : 6.871842709362768;
+        const inchesPerPixel = 1 / pixelsPerInch;
+        const dotsPerPixel = inchesPerPixel * DPI;
+ 
+        let pieceNum = 0;
+        patternPieces?.forEach(patternPiece => {
+            pieceNum++;
+            // Clone original piece and scale.
+            const originalPatternPiece = patternPiece.clone();
+            this._scaleAndRealign(originalPatternPiece, dotsPerPixel);
+
+            // Calculate the number of pages in a row and in a column
+            const boundBox = originalPatternPiece.getBoundingBox();
+            const numPagesX = Math.ceil(boundBox.width / pageSizeX);
+            const numPagesY = Math.ceil(boundBox.height / pageSizeY);
+            
+            let positionX = 0;
+            let positionY = 0;
+            for (let x = 0; x < numPagesX; x++) {
+                positionX = x * pageSizeX;
+            
+                for (let y = 0; y < numPagesY; y++) {
+                    positionY = y * pageSizeY;
+                    this.doc?.addPage();
+                    console.log("page num ", this.doc?.getNumberOfPages());
+
+                    if (!ctx) {
+                        throw new Error("Could not create 2D context for canvas.");
+                    }
```

Comment:
Can we move this check out of the double loop to line 79?

Reply:
Yes, thanks! I was making a new canvas for every piece, but I changed it to only creating the canvas for pdf once and forgot to move this out.